gnucash/gnome/assistant-stock-transaction.cpp

```diff
@@ -0,0 +1,1148 @@
+    // {
+    //     const char *memo = gtk_entry_get_text (GTK_ENTRY (info->cash_memo_edit));
+
+    //     /* asset split */
+    //     // account = gnc_tree_view_account_get_selected_account
+    //     //     (GNC_TREE_VIEW_ACCOUNT(info->asset_tree));
+    //     account_commits.emplace_back (account);
+
+    //     split = xaccMallocSplit (gnc_get_current_book ());
+    //     xaccAccountBeginEdit (account);
+    //     xaccSplitSetAccount (split, account);
+    //     xaccSplitSetParent (split, trans);
+    //     xaccSplitSetAmount (split, amount);
+    //     xaccSplitSetValue (split, amount);
+    //     xaccSplitSetMemo (split, memo);
+
+    //     /* income split */
+    //     // account = gnc_tree_view_account_get_selected_account
+    //     //     (GNC_TREE_VIEW_ACCOUNT(info->income_tree));
+    //     account_commits.emplace_back (account);
+
+    //     split = xaccMallocSplit (gnc_get_current_book ());
+    //     xaccAccountBeginEdit (account);
+    //     xaccSplitSetAccount (split, account);
+    //     xaccSplitSetParent (split, trans);
+    //     xaccSplitSetAmount (split, gnc_numeric_neg (amount));
+    //     xaccSplitSetValue (split, gnc_numeric_neg (amount));
+    //     xaccSplitSetMemo (split, memo);
+    // }
+
+    xaccTransCommitEdit (trans);
+
+    for (auto& it : account_commits)
+        xaccAccountCommitEdit (it);
+
+    gnc_resume_gui_refresh ();
+
+    gnc_close_gui_component_by_data (ASSISTANT_STOCK_TRANSACTION_CM_CLASS, info);
+}
+
+
+void
+stock_assistant_cancel (GtkAssistant *assistant, gpointer user_data)
+{
+    StockTransactionInfo *info = (StockTransactionInfo *)user_data;
+    gnc_close_gui_component_by_data (ASSISTANT_STOCK_TRANSACTION_CM_CLASS, info);
+}
+
+static GtkWidget * get_widget (GtkBuilder *builder, const gchar * ID)
+{
+    g_return_val_if_fail (builder && ID, nullptr);
+    auto obj = gtk_builder_get_object (builder, ID);
+    if (!obj)
+        g_warning ("get_widget ID '%s' not found. it may be a typo?", ID);
+    return GTK_WIDGET (obj);
+}
+
+static GtkWidget * create_gas (GtkBuilder *builder, gint row,
+                               std::vector<GNCAccountType> type, gnc_commodity *currency,
+                               const gchar *table_ID, const gchar *label_ID)
+{
+    auto table = get_widget (builder, table_ID);
+    auto label = get_widget (builder, label_ID);
+    auto gas = gnc_account_sel_new ();
+    GList *acct_list = nullptr;
+    for (auto& it : type)
+        acct_list = g_list_prepend (acct_list, (gpointer)it);
+    auto curr_list = g_list_prepend (nullptr, currency);
+    gnc_account_sel_set_new_account_ability (GNC_ACCOUNT_SEL (gas), true);
+    gnc_account_sel_set_acct_filters (GNC_ACCOUNT_SEL (gas), acct_list, curr_list);
+    gtk_widget_show (gas);
+    gtk_grid_attach (GTK_GRID(table), gas, 1, row, 1, 1);
+    gtk_label_set_mnemonic_widget (GTK_LABEL(label), gas);
+    g_list_free (acct_list);
+    g_list_free (curr_list);
+    return gas;
+}
```

Comment:
This one leaks `curr_list` however I can't figure out why. It's freed in #L955.
```
==38937== 24 bytes in 1 blocks are definitely lost in loss record 6,103 of 19,428
==38937==    at 0x4843839: malloc (in /usr/libexec/valgrind/vgpreload_memcheck-amd64-linux.so)
==38937==    by 0x4C3C088: g_malloc (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.6800.4)
==38937==    by 0x4C53374: g_slice_alloc (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.6800.4)
==38937==    by 0x4C2DF5A: g_list_copy_deep (in /usr/lib/x86_64-linux-gnu/libglib-2.0.so.0.6800.4)
==38937==    by 0x4F9A334: gnc_account_sel_set_acct_filters (gnc-account-sel.c:388)
==38937==    by 0x4D93E9E: create_gas(_GtkBuilder*, int, std::vector<GNCAccountType, std::allocator<GNCAccountType> >, gnc_commodity_s*, char const*, char const*) (assistant-stock-transaction.cpp:950)
==38937==    by 0x4D94C4E: stock_assistant_create(StockTransactionInfo*) (assistant-stock-transaction.cpp:1087)
==38937==    by 0x4D95030: gnc_stock_transaction_assistant (assistant-stock-transaction.cpp:1139)
==38937==    by 0x4E2C5C3: gnc_plugin_page_register_cmd_stock_assistant (gnc-plugin-page-register.c:4581)
==38937==    by 0x5A8CC0E: g_closure_invoke (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937==    by 0x5AA8EA5: ??? (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937==    by 0x5AAA883: g_signal_emit_valist (in /usr/lib/x86_64-linux-gnu/libgobject-2.0.so.0.6800.4)
==38937== 
```

Reply:
It's not leaking `curr_list`, it's leaking the *copy* of `curr_list` in `gas` because [`gnc_account_sel_finalize`](https://github.com/Gnucash/gnucash/blob/7e299cb2b0f2d586462b8115440b3d17ff2b90dd/gnucash/gnome-utils/gnc-account-sel.c#L394) only frees `acctTypeFilters`.

Reply:
> It's not leaking `curr_list`, it's leaking the _copy_ of `curr_list` in `gas` because [`gnc_account_sel_finalize`](https://github.com/Gnucash/gnucash/blob/7e299cb2b0f2d586462b8115440b3d17ff2b90dd/gnucash/gnome-utils/gnc-account-sel.c#L394) only frees `acctTypeFilters`.

This was a _sneaky_ leak...